### PR TITLE
Choose a better key format, cache app account hashes

### PR DESCRIPTION
**Pull request by jannotti**

### CODE REVIEW

File: ledger/internal/applications.go [hunk 1]
```diff
@@ -226,13 +226,13 @@ func (cs *roundCowState) NewBox(appIdx basics.AppIndex, key string, size uint64)
 		return fmt.Errorf("box %s exists for %d", key, appIdx)
 	}
 
-	record, err := cs.Get(appIdx.Address(), false)
+	record, err := cs.Get(appAddr, false)
```
Comment (ahangsu) [inline ledger/internal/applications.go:229]: one dumb question, why we can previously infer application address from appidx, but now we have to pass in `appAddr`? Does the previous implementation need to be corrected?

**Review by ahangsu — COMMENTED** [review]
took a quick pass

File: ledger/internal/applications.go (hunk 1 above)
Comment (jannotti) [reply ledger/internal/applications.go:229]: This is an optimization.  `appIdx.Address()` computes a hash.  There is a cache of such things on the calling side, so we pass it in instead.  The caller _must_ supply the address that corresponds to this app, or the MBR will get charged to the wrong account.

File: data/transactions/logic/box.go [hunk 2]
```diff
@@ -149,19 +152,22 @@ func opBoxDel(cx *EvalContext) error {
...
+	   the boxes of a certain app with a certain prefix.
+
+	   The "bx:" prefix is so that the kvstore might be usable for things
+	   besides boxes.
+	*/
+	key := make([]byte, 3 /* bx: */ +8 /* appIdx, big-endian */ +len(name))
+	copy(key, "bx:")
+	binary.BigEndian.PutUint64(key[3:], uint64(appIdx))
+	copy(key[11:], name)
+	return string(key)
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/box.go:172]: I just didn't want to convert app id to ascii every time. I kind of like the fixed length, but I don't think it's too bif of a deal.

File: ledger/internal/applications.go (hunk 1 above)
Comment (ahangsu) [reply ledger/internal/applications.go:229]: I see, so what we know is that caller maintains a cache containing such address, so we do not need to re-compute.

Another thing, what is MBR standing for?

**Review by michaeldiamant — APPROVED** [review]
@jannotti These changes make sense to me - I'll aim to be more cognizant of `appIdx.Address()` invocations in performance sensitive code paths.

Aside:  I briefly considered if it'd be helpful to define a variant of `AppIndex` that does _not_ expose `Address`.  Such a data type prevents unintended, expensive `Address` generation without painfully obvious gymnastics.  Though such a change is definitely beyond the PR's scope.

**Pull request merged**